### PR TITLE
Dm 317 implement interface register login

### PR DESCRIPTION
**Pull request by jenniferboedker**

### CODE REVIEW

**Comment by jenniferboedker** [comment]
There seems to be an issue with the register layout and I cant find out what it is, maybe you have an idea?

File: webapp/src/main/java/life/qbic/Application.java [hunk 1]
```diff
@@ -23,10 +19,10 @@
 @PWA(name = "Data Manager", shortName = "Data Manager", offlineResources = {"images/logo.png"})
 @NpmPackage(value = "line-awesome", version = "1.3.0")
 public class Application extends SpringBootServletInitializer implements AppShellConfigurator {
-
   public static void main(String[] args) {
-    ConfigurableApplicationContext context = SpringApplication.run(Application.class, args);
-    var userJpa = context.getBean(UserJpaRepository.class);
+    SpringApplication.run(Application.class, args);
+
+    /*var userJpa = context.getBean(UserJpaRepository.class);
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/Application.java:25]: Unused code can be removed

File: domain/src/main/java/life/qbic/usermanagement/User.java [hunk 2]
```diff
@@ -145,6 +146,15 @@ public String getEmail() {
     return this.email;
   }
 
+  /**
+   * Checks if a given password is correct for a user
+   * @param password Password that is being validated
+   * @return true, if the given password is correct for the user
+   */
+  public Boolean checkPassword(String password) {
```
Comment (sven1103) [inline domain/src/main/java/life/qbic/usermanagement/User.java:154]: I would prefer not to pass clear text passwords around in the application. Instead, pass the encrypted password as argument.

You can get the encrypted password hash from the User object you bind the input data from the form :)

File: webapp/src/main/java/life/qbic/security/AuthenticatedUser.java [hunk 3]
```diff
@@ -16,17 +16,18 @@
...
         SecurityContext context = SecurityContextHolder.getContext();
         return Optional.ofNullable(context.getAuthentication())
                 .filter(authentication -> !(authentication instanceof AnonymousAuthenticationToken));
     }
 
-    public Optional<TestUser> get() {
-        return getAuthentication().map(authentication -> userRepository.findByUsername(authentication.getName()));
+    public Optional<User> get() {
+        //return getAuthentication().map(authentication -> userRepository.findByUsername(authentication.getName()));
+        return null; //todo
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/security/AuthenticatedUser.java:30]: Can you add a more detailed TODO statement here please, so we know what needs to be done

File: webapp/src/main/java/life/qbic/security/UserDetailsServiceImpl.java [hunk 4]
```diff
@@ -15,24 +15,27 @@
...
-                    getAuthorities(testUser));
-        }
+        //User testUser = userRepository.findByUsername(username);
+        //if (testUser == null) {
+        //    throw new UsernameNotFoundException("No testUser present with username: " + username);
+        //} else {
+            //todo fix me
+            //return new org.springframework.security.core.userdetails.User(testUser.getUsername(), testUser.getHashedPassword(),
+            //        getAuthorities(testUser));
+            return null;
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/security/UserDetailsServiceImpl.java:30]: commented out code can be removed, and a TODO statement addded

File: webapp/src/main/java/life/qbic/views/MainController.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            mainLayout.login.getUI().ifPresent(ui ->
+                    ui.navigate("login"));
+
+        });
+
+        mainLayout.register.addClickListener(event -> {
+            mainLayout.register.getUI().ifPresent(ui ->
+                    ui.navigate("register"));
+        });
+    }
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/MainController.java:28]: Should be a handler, not a controller, according to our architecture design

File: webapp/src/main/java/life/qbic/views/MainLayout.java [hunk 6]
```diff
@@ -2,142 +2,69 @@
...
-    }
-
-    private AuthenticatedUser authenticatedUser;
-    private AccessAnnotationChecker accessChecker;
-
-    public MainLayout(AuthenticatedUser authenticatedUser, AccessAnnotationChecker accessChecker) {
-        this.authenticatedUser = authenticatedUser;
-        this.accessChecker = accessChecker;
+    protected Button register;
+    protected Button login;
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/MainLayout.java:22]: can be public, there is no harm in exposing the elements of the layout

File: webapp/src/main/java/life/qbic/views/MainLayout.java [hunk 7]
```diff
@@ -2,142 +2,69 @@
...
-
-        H1 appName = new H1("Data Manager");
-        appName.addClassNames("my-0", "me-auto", "text-l");
-        layout.add(appName);
-
-        Optional<TestUser> maybeUser = authenticatedUser.get();
-        if (maybeUser.isPresent()) {
-            TestUser testUser = maybeUser.get();
+        HorizontalLayout header = createHeaderLayout();
+        HorizontalLayout buttons = createHeaderButtonLayout();
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/MainLayout.java:30]: Java 17 way for local variables:
```suggestion
        var header = createHeaderLayout();
        var buttons = createHeaderButtonLayout();
```

File: webapp/src/main/java/life/qbic/views/MainLayout.java [hunk 8]
```diff
@@ -2,142 +2,69 @@
...
-            name.addClassNames("font-medium", "text-s", "text-secondary");
+    private HorizontalLayout createHeaderLayout() {
+        H1 appName = styleHeaderTitle();
 
-            layout.add(avatar, name);
-        } else {
-            Anchor loginLink = new Anchor("login", "Sign in");
-            layout.add(loginLink);
-        }
+        HorizontalLayout header = new HorizontalLayout(appName);
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/MainLayout.java:40]:
```suggestion
        var header = new HorizontalLayout(appName);
```

File: webapp/src/main/java/life/qbic/views/MainLayout.java [hunk 9]
```diff
@@ -2,142 +2,69 @@
...
-    private MenuItemInfo[] createMenuItems() {
-        return new MenuItemInfo[]{ //
-                new MenuItemInfo("Hello World", "la la-globe", HelloWorldView.class), //
-
-                new MenuItemInfo("About", "la la-file", AboutView.class), //
+    private void styleHeaderButtons() {
+        register = new Button("Register");
 
-        };
+        login = new Button("Login");
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/MainLayout.java:66]: style vs. creatation. A function should do only one thing.

File: webapp/src/main/java/life/qbic/views/login/LoginHandler.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    LoginHandler(@Autowired UserJpaRepository repository){
+        this.userRepository = repository;
+    }
+
+    @Override
+    public boolean register(LoginLayout loginView) {
+        if (registeredLoginView != loginView) {
+            this.registeredLoginView = loginView;
+            // orchestrate view
+            addListener();
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/login/LoginHandler.java:25]:
```suggestion
            registeredLoginView = loginView;
            // orchestrate view
            addListener();
```

or 

```suggestion
            this.registeredLoginView = loginView;
            // orchestrate view
            this.addListener();
```

to be consistent

File: webapp/src/main/java/life/qbic/views/login/LoginHandler.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+                users.get(0).checkPassword(registeredLoginView.password.getValue());
+                //todo authorization: show the correct route now --> security context
+                UI.getCurrent().navigate("about"); //could be dashboard later
+                Notification.show("It worked");
+            }catch (RuntimeException r){
+                //todo show error in ui
+                Notification.show("Wrong email or password");
+            }
+        });
+    }
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/login/LoginHandler.java:46]: You can use &#x2325; &#x2318; L to format (best to get it in automated for your left hand, to always reformat regularly without conciously doing it)

File: webapp/src/main/java/life/qbic/views/login/LoginHandlerInterface.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package life.qbic.views.login;
+
+/**
+ *
+ */
+public interface LoginHandlerInterface {
+
+  boolean register(LoginLayout loginView);
+
+}
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/login/LoginHandlerInterface.java:10]: Description and purpose missing

File: webapp/src/main/java/life/qbic/views/login/LoginLayout.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+public class LoginLayout extends VerticalLayout {
+
+    protected EmailField email;
+
+    protected PasswordField password;
+
+    protected Button loginButton;
+
+    protected Span registerSpan;
+
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/login/LoginLayout.java:41]: can be public

File: webapp/src/main/java/life/qbic/views/login/LoginLayout.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    private void styleLoginButton() {
+        loginButton = new Button("Login");
+        loginButton.addThemeVariants(ButtonVariant.LUMO_PRIMARY);
+        loginButton.setEnabled(true);
+        loginButton.setWidthFull();
+    }
+
+    private void styleEmailField(){
+        email = new EmailField("Email");
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/login/LoginLayout.java:97]: a function shall do only one thing

File: webapp/src/main/java/life/qbic/views/register/RegisterHandler.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    return false;
+  }
+
+  private void addListener() {
+    registeredRegisterLayout.registerButton.addClickListener(
+        event -> {
+          List<User> user =
+              userJpaRepository.findUsersByEmail(registeredRegisterLayout.email.getValue());
+
+          if (user.size() > 0) {
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/register/RegisterHandler.java:47]: I think here we should go into the model, and not execute this kind of logic in the UI. 

Instead, i suggest to call the use case, make the check there and execute call back methodes via the presenter.

File: webapp/src/main/java/life/qbic/views/register/RegisterHandler.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+            Notification.show("Email is already in use!");
+          }
+          try {
+            String password = registeredRegisterLayout.password.getValue();
+            String email = registeredRegisterLayout.email.getValue();
+            String name = registeredRegisterLayout.fullName.getValue();
+
+            User newUser = User.create(password, name, email);
+            userJpaRepository.save(newUser);
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/register/RegisterHandler.java:56]: same here, that is application logic and should not be in the UI layer

File: webapp/src/main/java/life/qbic/views/register/RegisterLayout.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+    protected EmailField email;
+
+    protected PasswordField password;
+
+    protected TextField fullName;
+
+    protected Button registerButton;
+
+    protected Span loginSpan;
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/register/RegisterLayout.java:39]: can be public

File: webapp/src/main/java/life/qbic/views/register/RegisterHandlerInterface.java [hunk 18]
```diff
@@ -0,0 +1,7 @@
+package life.qbic.views.register;
+
+
+public interface RegisterHandlerInterface {
+
+    boolean register(RegisterLayout registerLayout);
+}
```
Comment (sven1103) [inline webapp/src/main/java/life/qbic/views/register/RegisterHandlerInterface.java:7]: missing description and purpose

**Review by sven1103 — CHANGES_REQUESTED** [review]
Dear Jenny, thanks a lot for that, it already looks really awesome.

Technical comments are there, I have also some behavioural feedback:

1. Register Button should be not enabled, if the input is not valid. 
2. The notification `not a valid password` is irritating. So either we are more specific, like to short, of we just make the input field red.
3. IMO the fields are a little bit to short, especially for the email address, which gets cut off
4. I think the enter key on the keyboard should lead to the button activation (register or login). I always use it and it did not work
5. The notifications from the app are on the bottom left
6. If any exception occurs during the registration, the user should stay on the registration view and the email field been invalidated

File: domain/src/main/java/life/qbic/usermanagement/User.java (hunk 2 above)
Comment (jenniferboedker) [reply domain/src/main/java/life/qbic/usermanagement/User.java:154]: But that's in the user class.. I need to create

File: webapp/src/main/java/life/qbic/views/register/RegisterHandler.java (hunk 16 above)
Comment (jenniferboedker) [reply webapp/src/main/java/life/qbic/views/register/RegisterHandler.java]: tackled by #8

**Comment by jenniferboedker** [comment]
> Dear Jenny, thanks a lot for that, it already looks really awesome.
> 
> Technical comments are there, I have also some behavioural feedback:
> 
> 1. Register Button should be not enabled, if the input is not valid.
As discussed, this is commen to do so. Therefore, I left it as is.
> 2. The notification `not a valid password` is irritating. So either we are more specific, like to short, of we just make the input field red.
I removed the text since we have specifications for the textfield 
> 3. IMO the fields are a little bit to short, especially for the email address, which gets cut off
TODO
> 4. I think the enter key on the keyboard should lead to the button activation (register or login). I always use it and it did not work
TODO
> 5. The notifications from the app are on the bottom left
This should be tackled by another PR, its the styling of the notifications
> 6. If any exception occurs during the registration, the user should stay on the registration view and the email field been invalidated
What do mean with invalidated and why the email field?

**Review by sven1103 — APPROVED** [review]

**Pull request merged**